### PR TITLE
Update setlogurl method for GoLocal support

### PR DESCRIPTION
**Pull request by Cloudwu1**
## Description
Change start API to take overridden log URL for One Collector
Add info message when setLogUrl is called and overridden.
Update setLogUrl method to change Ingestion endpoint immediately
If appSecret exists => change App Center Ingestion endpoint.
If appSecret doesn't exist => change One Collector Ingestion endpoint.

### CODE REVIEW

**Comment by jaelim-ms** [comment]
After review, we will convert this branch to feature branch. Please do not merge

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 1]
```diff
@@ -494,7 +501,16 @@ private synchronized void setInstanceLogUrl(final String logUrl) {
 
                 @Override
                 public void run() {
-                    mChannel.setLogUrl(logUrl);
+                    if (mChannel != null) {
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: mChannel is not null, please remove if, 

Some changes need to be done in AppCenter to delay setLogUrl in the channel to when we start. Keep logurl in appcenter class if called before start and apply it at start.

Comment (Cloudwu1) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: Resolved in new commit.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 2]
```diff
@@ -494,7 +501,14 @@ private synchronized void setInstanceLogUrl(final String logUrl) {
 
                 @Override
                 public void run() {
-                    mChannel.setLogUrl(logUrl);
+                        if (mAppSecret != null) {
+                            AppCenterLog.info(LOG_TAG,"The log url of App Center endpoint was changed to " + logUrl);
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]:
```suggestion
                            AppCenterLog.info(LOG_TAG, "The log url of App Center endpoint was changed to " + logUrl);
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 3]
```diff
@@ -494,7 +501,14 @@ private synchronized void setInstanceLogUrl(final String logUrl) {
...
+                        } else {
+                            AppCenterLog.info(LOG_TAG,"The log url of One Collector endpoint was changed to " + logUrl);
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]:
```suggestion
                            AppCenterLog.info(LOG_TAG, "The log url of One Collector endpoint was changed to " + logUrl);
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 4]
```diff
@@ -744,7 +758,12 @@ private void finishConfiguration(boolean configureFromApp) {
         mLogSerializer = new DefaultLogSerializer();
         mLogSerializer.addLogFactory(StartServiceLog.TYPE, new StartServiceLogFactory());
         mLogSerializer.addLogFactory(CustomPropertiesLog.TYPE, new CustomPropertiesLogFactory());
-        mChannel = new DefaultChannel(mApplication, mAppSecret, mLogSerializer, mHandler);
+        if(mAppSecret != null){
+            mChannel = new DefaultChannel(mApplication, mAppSecret, mLogSerializer, mHandler);
+        }else{
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]:
```suggestion
        } else {
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 5]
```diff
@@ -744,7 +758,12 @@ private void finishConfiguration(boolean configureFromApp) {
         mLogSerializer = new DefaultLogSerializer();
         mLogSerializer.addLogFactory(StartServiceLog.TYPE, new StartServiceLogFactory());
         mLogSerializer.addLogFactory(CustomPropertiesLog.TYPE, new CustomPropertiesLogFactory());
-        mChannel = new DefaultChannel(mApplication, mAppSecret, mLogSerializer, mHandler);
+        if(mAppSecret != null){
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]:
```suggestion
        if (mAppSecret != null) {
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 6]
```diff
@@ -744,7 +758,12 @@ private void finishConfiguration(boolean configureFromApp) {
         mLogSerializer = new DefaultLogSerializer();
         mLogSerializer.addLogFactory(StartServiceLog.TYPE, new StartServiceLogFactory());
         mLogSerializer.addLogFactory(CustomPropertiesLog.TYPE, new CustomPropertiesLogFactory());
-        mChannel = new DefaultChannel(mApplication, mAppSecret, mLogSerializer, mHandler);
+        if (mAppSecret != null) {
+            mChannel = new DefaultChannel(mApplication, mAppSecret, mLogSerializer, mHandler);
+        } else {
+            mIngestion = new OneCollectorIngestion(mApplication, mLogSerializer);
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: You must not change when One Collector is instantiated.

Instead you must keep a reference to the log url and apply setLogUrl when one collector is created, also remove the if and roll back this section of the code.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 7]
```diff
@@ -180,6 +182,11 @@
      */
     private AppCenterHandler mAppCenterHandler;
 
+    /**
+     * The ingestion object used to send batches to the server.
+     */
+    private Ingestion mIngestion;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: remove that and replace by mLogUrl, see other comment.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java [hunk 8]
```diff
@@ -126,6 +126,10 @@ public DefaultChannel(@NonNull Context context, String appSecret, @NonNull LogSe
         this(context, appSecret, buildDefaultPersistence(context, logSerializer), new AppCenterIngestion(context, logSerializer), appCenterHandler);
     }
 
+    public DefaultChannel(@NonNull Context context, String appSecret, @NonNull LogSerializer logSerializer, @NonNull Ingestion ingestion, @NonNull Handler appCenterHandler) {
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java]: please also roll back this one

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 7 above)
Comment (Cloudwu1) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: Resolve in new commit.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 9]
```diff
@@ -775,10 +772,14 @@ private void finishConfiguration(boolean configureFromApp) {
         }
         mChannel.setEnabled(enabled);
         mChannel.addGroup(CORE_GROUP, DEFAULT_TRIGGER_COUNT, DEFAULT_TRIGGER_INTERVAL, DEFAULT_TRIGGER_MAX_PARALLEL_REQUESTS, null, null);
-        if (mLogUrl != null) {
+        if (mLogUrl != null && mAppSecret != null) {
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: you can move this to the other if to combine the checks and code to simplify

```
if (mLogUrl != null) {
    if (mAppSecret != null) {
            mChannel.setLogUrl(mLogUrl);
    }
    else {
           mOneCollectorChannelListener.setLogUrl(mLogUrl);
    }
}
```

In line 779

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 10]
```diff
@@ -775,10 +772,14 @@ private void finishConfiguration(boolean configureFromApp) {
...
             mChannel.setLogUrl(mLogUrl);
         }
-        mChannel.addListener(new OneCollectorChannelListener(mApplication, mChannel, mLogSerializer, IdHelper.getInstallId()));
+        mOneCollectorChannelListener = new OneCollectorChannelListener(mApplication, mChannel, mLogSerializer, IdHelper.getInstallId());
+        if(mLogUrl != null && mAppSecret == null){
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: reformat file, also please commit via Android Studio with these options:

Screen Shot 2019-03-12 at 11 52 09 AM

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 11]
```diff
@@ -190,6 +190,11 @@
      */
     private DefaultAppCenterFuture<Boolean> mSetMaxStorageSizeFuture;
 
+    /**
+     *
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: Missing description.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/OneCollectorChannelListener.java [hunk 12]
```diff
@@ -89,6 +89,10 @@ public OneCollectorChannelListener(@NonNull Context context, @NonNull Channel ch
         mIngestion = new OneCollectorIngestion(context, mLogSerializer);
     }
 
+    public void setLogUrl(@NonNull String logUrl) {
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/OneCollectorChannelListener.java]: Add Javadoc

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 13]
```diff
@@ -850,6 +850,26 @@ public void setLogUrl() {
         verify(mChannel).setLogUrl(logUrl);
     }
 
+    @Test
+    public void setLogUrlWithNoAppSecret() {
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: Good but need more testing.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 10 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: @guperrot Outdated

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 11 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: Yep, this is just placeholder/reminder to add 😄

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java (hunk 13 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: Sure, CI clearly says it

**Comment by guperrot** [comment]
Tested all launch modes and also setLogUrl after start in Sasquatch: behavior OK.

**Comment by guperrot** [comment]
> ## Description
> Change start API to take overridden log URL for One Collector
> Add info message when setLogUrl is called and overridden.
> Update setLogUrl method to change Ingestion endpoint immediately
> If appSecret exists => change App Center Ingestion endpoint.
> If appSecret doesn't exist => change One Collector Ingestion endpoint.

Please don't erase the template next time, and check the boxes that apply.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 14]
```diff
@@ -850,6 +849,28 @@ public void setLogUrl() {
         verify(mChannel).setLogUrl(logUrl);
     }
 
+    @Test
+    public void setLogUrlWithNoAppSecret() throws Exception {
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: With target token should be in the name.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 15]
```diff
@@ -850,6 +849,28 @@ public void setLogUrl() {
...
+
+        /* Change log URL before start. */
+        String logUrl = "http://mock";
+        AppCenter.setLogUrl(logUrl);
+
+        /* No effect for now. */
+        verify(listener, never()).setLogUrl(logUrl);
+
+        /* Start should propagate the log URL without App Secret. */
+        AppCenter.start(mApplication, DUMMY_TARGET_TOKEN_STRING, DummyService.class);
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: Maybe make 2 variants of this test (sharing most of the code) as this works also with no secret at all.

**Comment by guperrot** [comment]
@MatkovIvan I let you merge so that you can see what I pushed.

**Pull request merged**